From acf97e320f51fabdfe9e82f160698e0d4b66892d Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 1 Mar 2024 11:12:58 +0100
Subject: [PATCH 116/217] net: dovetail: move oob context flags to device
 private flags

Out-of-band capability is a generic information which should be
advertised by the device directly.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/linux/netdevice.h | 61 ++++++++++++++++++++++++++++++++-------
 include/net/netoob.h      |  6 ----
 2 files changed, 51 insertions(+), 16 deletions(-)

diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 2572745bf..5e778a36b 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -1364,14 +1364,12 @@ struct netdev_net_notifier {
  * 					dma_addr_t *dma_addr);
  *	Allocate a buffer for out-of-band I/O. The memory must be immediately
  *	DMA-suitable, living in the linear kernel address space. i.e. no highmem.
- *	This handler must be provided if IFF_OOB_CAPABLE is set in the out-of-band
- *	context flags.
+ *	This handler must be provided if dev->oob_capable is set.
  * void	(*ndo_free_oob_skb)(struct net_device *dev,
  *			    struct sk_buff *skb,
  *			    dma_addr_t dma_addr);
  *	Free a buffer previously allocated from ndo_alloc_oob_skb() exclusively.
- *	This handler must be provided if IFF_OOB_CAPABLE is set in the out-of-band
- *	context flags.
+ *	This handler must be provided if dev->oob_capable is set.
  */
 struct net_device_ops {
 	int			(*ndo_init)(struct net_device *dev);
@@ -1710,6 +1708,38 @@ enum netdev_priv_flags {
 	IFF_TX_SKB_NO_LINEAR		= BIT_ULL(31),
 };
 
+#define IFF_802_1Q_VLAN			IFF_802_1Q_VLAN
+#define IFF_EBRIDGE			IFF_EBRIDGE
+#define IFF_BONDING			IFF_BONDING
+#define IFF_ISATAP			IFF_ISATAP
+#define IFF_WAN_HDLC			IFF_WAN_HDLC
+#define IFF_XMIT_DST_RELEASE		IFF_XMIT_DST_RELEASE
+#define IFF_DONT_BRIDGE			IFF_DONT_BRIDGE
+#define IFF_DISABLE_NETPOLL		IFF_DISABLE_NETPOLL
+#define IFF_MACVLAN_PORT		IFF_MACVLAN_PORT
+#define IFF_BRIDGE_PORT			IFF_BRIDGE_PORT
+#define IFF_OVS_DATAPATH		IFF_OVS_DATAPATH
+#define IFF_TX_SKB_SHARING		IFF_TX_SKB_SHARING
+#define IFF_UNICAST_FLT			IFF_UNICAST_FLT
+#define IFF_TEAM_PORT			IFF_TEAM_PORT
+#define IFF_SUPP_NOFCS			IFF_SUPP_NOFCS
+#define IFF_LIVE_ADDR_CHANGE		IFF_LIVE_ADDR_CHANGE
+#define IFF_MACVLAN			IFF_MACVLAN
+#define IFF_XMIT_DST_RELEASE_PERM	IFF_XMIT_DST_RELEASE_PERM
+#define IFF_L3MDEV_MASTER		IFF_L3MDEV_MASTER
+#define IFF_NO_QUEUE			IFF_NO_QUEUE
+#define IFF_OPENVSWITCH			IFF_OPENVSWITCH
+#define IFF_L3MDEV_SLAVE		IFF_L3MDEV_SLAVE
+#define IFF_TEAM			IFF_TEAM
+#define IFF_RXFH_CONFIGURED		IFF_RXFH_CONFIGURED
+#define IFF_PHONY_HEADROOM		IFF_PHONY_HEADROOM
+#define IFF_MACSEC			IFF_MACSEC
+#define IFF_NO_RX_HANDLER		IFF_NO_RX_HANDLER
+#define IFF_FAILOVER			IFF_FAILOVER
+#define IFF_FAILOVER_SLAVE		IFF_FAILOVER_SLAVE
+#define IFF_L3MDEV_RX_HANDLER		IFF_L3MDEV_RX_HANDLER
+#define IFF_TX_SKB_NO_LINEAR		IFF_TX_SKB_NO_LINEAR
+
 /* Specifies the type of the struct net_device::ml_priv pointer */
 enum netdev_ml_priv_type {
 	ML_PRIV_NONE,
@@ -2390,6 +2420,8 @@ struct net_device {
 	unsigned long		change_proto_down:1;
 	unsigned long		netns_local:1;
 	unsigned long		fcoe_mtu:1;
+	unsigned long		oob_capable:1;
+	unsigned long		oob_port:1;
 
 	struct list_head	net_notifier_list;
 
@@ -4353,24 +4385,29 @@ static inline void netif_disable_oob_diversion(struct net_device *dev)
 
 int netif_xmit_oob(struct sk_buff *skb);
 
+static inline void netdev_set_oob_capable(struct net_device *dev)
+{
+	dev->oob_capable = 1;
+}
+
 static inline bool netdev_is_oob_capable(struct net_device *dev)
 {
-	return !!(dev->oob_context.flags & IFF_OOB_CAPABLE);
+	return dev->oob_capable;
 }
 
 static inline void netdev_enable_oob_port(struct net_device *dev)
 {
-	dev->oob_context.flags |= IFF_OOB_PORT;
+	dev->oob_port = 1;
 }
 
 static inline void netdev_disable_oob_port(struct net_device *dev)
 {
-	dev->oob_context.flags &= ~IFF_OOB_PORT;
+	dev->oob_port = 0;
 }
 
 static inline bool netdev_is_oob_port(struct net_device *dev)
 {
-	return !!(dev->oob_context.flags & IFF_OOB_PORT);
+	return dev->oob_port;
 }
 
 static inline struct sk_buff *netdev_alloc_oob_skb(struct net_device *dev,
@@ -4388,7 +4425,11 @@ static inline void netdev_free_oob_skb(struct net_device *dev,
 
 #else
 
-static inline bool netif_receive_oob(struct sk_buff *skb)
+static inline void netdev_set_oob_capable(struct net_device *dev)
+{
+}
+
+static inline bool netdev_is_oob_capable(struct net_device *dev)
 {
 	return false;
 }
@@ -4398,7 +4439,7 @@ static inline bool netif_oob_diversion(const struct net_device *dev)
 	return false;
 }
 
-static inline bool netdev_is_oob_capable(struct net_device *dev)
+static inline bool netif_receive_oob(struct sk_buff *skb)
 {
 	return false;
 }
diff --git a/include/net/netoob.h b/include/net/netoob.h
index 55fd2357a..02ec893d7 100644
--- a/include/net/netoob.h
+++ b/include/net/netoob.h
@@ -5,14 +5,8 @@
 #include <dovetail/net.h>
 #include <dovetail/netdevice.h>
 
-/* Device supports direct out-of-band operations (RX & TX) */
-#define IFF_OOB_CAPABLE		BIT(0)
-/* Device is an out-of-band port */
-#define IFF_OOB_PORT		BIT(1)
-
 /* Netdevice context extension for oob handling. */
 struct oob_netdev_context {
-	int flags;
 	struct oob_netdev_state dev_state;
 };
 
-- 
2.30.2

